plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -159,6 +168,58 @@ func (g *genericScheduler) selectHost(priorityList schedulerapi.HostPriorityList
 	return priorityList[ix].Host, nil
 }
 
+// preempt finds nodes with pods that can be preempted to make room for "pod" to
+// schedule. It chooses one of the nodes and preempts the pods on the node and
+// returns the node and the list of preempted pods if such a node is found.
+func (g *genericScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister, scheduleErr error) (*v1.Node, []*v1.Pod, error) {
+	// Scheduler may return various types of errors. Consider preemption only if
+	// the error is of type FitError.
+	fitError, ok := scheduleErr.(*FitError)
+	if !ok || fitError == nil {
+		return nil, nil, nil
+	}
+	err := g.cache.UpdateNodeNameToInfoMap(g.cachedNodeInfoMap)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !podEligibleToPreemptOthers(pod, g.cachedNodeInfoMap) {
```

Comment:
I started wondering about this.
Do we somehow protect ourselves from the situation that there will be new pods coming all the time that will be smaller, and will effectively blocking our high priority pod from scheduling?

I'm talking for example about scanario like this:
- there is a single node,
- there is Deployment with 10 replicas (low priority) each of the replica with 30s grace period
- now a high priority pod is coming that requires the whole onde
- so we call preempt on all replicas
- but before last pod is removed, the first one is recreated by RS controller and since there is a place on that node scheduler puts it here
- when preempted pods are removed, there is again no place for the high priority pod
- so we start from the begining.

And at the end of the day, we pretty much starved the high-priority pod with low-priority pods.

Unless I'm missing something and this problem is not solved, we shouldn't rush for 1.8 with this change.

@davidopp

Reply:
@bsalamat Can you write a separate doc describing this problem, the corner case(s) that the current approach (in this PR) doesn't address, the plan we have for fixing it, and alternatives approaches for fixing it (e.g. mark the entire node as "can't schedule here" until all preemption victims have exited and preemptor has started, maybe also the Borg approach though it's super-complicated).

Reply:
Sure. I will write a doc.

Reply:
The `podEligibleToPreemptOthers` function already has a TODO. I added another one to `Preempt` as well.